File: app/schemas/reports.py
from datetime import date, datetime
from typing import Literal, Self

from pydantic import BaseModel, Field, model_validator

from app.schemas.performance import KpiTrendPoint


class EmployeeReportRequest(BaseModel):
    employee_id: str = Field(min_length=1, max_length=200)
    period_weeks: Literal[4, 8, 12] | None = None
    start_date: date | None = None
    end_date: date | None = None

    @model_validator(mode="after")
    def validate_period(self) -> Self:
        if self.period_weeks is not None and (
            self.start_date is not None or self.end_date is not None
        ):
            raise ValueError(
                "period_weeks cannot be combined with start_date or end_date."
            )
        if (self.start_date is None) != (self.end_date is None):
            raise ValueError("start_date and end_date must be provided together.")
        if (
            self.start_date is not None
            and self.end_date is not None
            and self.start_date > self.end_date
        ):
            raise ValueError("start_date must be on or before end_date.")
        return self


class ReportPeriod(BaseModel):
    start_date: date
    end_date: date
    prior_start_date: date | None = None
    prior_end_date: date | None = None


class ReportKpiSection(BaseModel):
    name: Literal["Productivity", "Compliance", "Quality"]
    score: float
    weight: float
    explanation: str


class ReportFinding(BaseModel):
    code: str
    severity: Literal["error", "warning", "info"]
    scoring_impact: Literal[
        "blocks_score",
        "excluded_from_scoring",
        "lowers_confidence",
        "affects_score",
        "none",
    ]
    message: str
    occurrence_count: int = Field(ge=1)
    record_ids: list[str]
    evidence_links: list[str]


class EmployeeReportData(BaseModel):
    employee_id: str
    employee_name: str | None
    team: str | None
    role: str | None
    period: ReportPeriod
    generated_at: datetime
    overall_score: float | None
    result_status: str
    performance_tier: str | None
    data_confidence: float
    confidence_threshold: float
    confidence_explanation: str
    kpis: list[ReportKpiSection] = Field(min_length=3, max_length=3)
    trends: list[KpiTrendPoint]
    prior_overall_score: float | None = None
    overall_score_change: float | None = None
    findings: list[ReportFinding]
    supporting_record_ids: list[str]
    metric_definitions: list[str]
    manager_review_notice: str


class EmployeeReportPreviewResponse(BaseModel):
    report: EmployeeReportData
    pdf_generated_by: Literal["browser"] = "browser"

File: web/src/types/reports.ts
import type { KpiTrendPoint } from '@/types/analysis'

export interface EmployeeReportRequest {
  employee_id: string
  start_date: string
  end_date: string
}

export interface ReportPeriod {
  start_date: string
  end_date: string
  prior_start_date: string | null
  prior_end_date: string | null
}

export interface ReportKpiSection {
  name: 'Productivity' | 'Compliance' | 'Quality'
  score: number
  weight: number
  explanation: string
}

export interface ReportFinding {
  code: string
  severity: 'error' | 'warning' | 'info'
  scoring_impact: string
  message: string
  occurrence_count: number
  record_ids: string[]
  evidence_links: string[]
}

export interface EmployeeReportData {
  employee_id: string
  employee_name: string | null
  team: string | null
  role: string | null
  period: ReportPeriod
  generated_at: string
  overall_score: number | null
  result_status: string
  performance_tier: string | null
  data_confidence: number
  confidence_threshold: number
  confidence_explanation: string
  kpis: ReportKpiSection[]
  trends: KpiTrendPoint[]
  prior_overall_score: number | null
  overall_score_change: number | null
  findings: ReportFinding[]
  supporting_record_ids: string[]
  metric_definitions: string[]
  manager_review_notice: string
}

export interface EmployeeReportPreviewResponse {
  report: EmployeeReportData
  pdf_generated_by: 'browser'
}
